src/solvers/refinement/string_constraint_generator_format.cpp

```diff
@@ -0,0 +1,424 @@
+/// String.format. Assumes the argument is a structured expression in which
+/// contains the fields: string expr, int, float, char, boolean, hashcode,
+/// date_time. The correct component will be fetched depending on the format
+/// specifier.
+/// \param fs: a format specifier
+/// \param arg: a struct containing the possible value of the argument to format
+/// \param ref_type: a type  for refined string type
+/// \return String expression representing the output of String.format.
+string_exprt string_constraint_generatort::add_axioms_for_format_specifier(
+  const format_specifiert &fs,
+  const struct_exprt &arg,
+  const refined_string_typet &ref_type)
+{
+  switch(fs.conversion)
+  {
+  case format_specifiert::DECIMAL_INTEGER:
+    return add_axioms_from_int(
+      get_component_in_struct(arg, ID_int), MAX_INTEGER_LENGTH, ref_type);
+  case format_specifiert::OCTAL_INTEGER:
+    // TODO: conversion of octal not implemented
+  case format_specifiert::HEXADECIMAL_INTEGER:
+    return add_axioms_from_int_hex(
+      get_component_in_struct(arg, ID_int), ref_type);
+  case format_specifiert::SCIENTIFIC:
+  case format_specifiert::GENERAL:
+  case format_specifiert::DECIMAL_FLOAT:
+  case format_specifiert::HEXADECIMAL_FLOAT:
+    return add_axioms_for_string_of_float(
+      get_component_in_struct(arg, ID_float), ref_type);
+  case format_specifiert::CHARACTER:
+    return add_axioms_from_char(
+      get_component_in_struct(arg, ID_char), ref_type);
+  case format_specifiert::BOOLEAN:
+    return add_axioms_from_bool(
+      get_component_in_struct(arg, ID_boolean), ref_type);
+  case format_specifiert::STRING:
+    return get_string_expr(get_component_in_struct(arg, "string_expr"));
+  case format_specifiert::HASHCODE:
+    // TODO: no implementation for hashcode
+    return add_axioms_from_int(
+      get_component_in_struct(arg, "hashcode"), MAX_INTEGER_LENGTH, ref_type);
+  case format_specifiert::LINE_SEPARATOR:
+    // TODO: the constant should depend on the system: System.lineSeparator()
+    return add_axioms_for_constant("\n", ref_type);
+  case format_specifiert::PERCENT_SIGN:
+    return add_axioms_for_constant("%", ref_type);
+  case format_specifiert::SCIENTIFIC_UPPER:
+  case format_specifiert::GENERAL_UPPER:
+  case format_specifiert::HEXADECIMAL_FLOAT_UPPER:
+  case format_specifiert::CHARACTER_UPPER:
+  case format_specifiert::DATE_TIME_UPPER:
+  case format_specifiert::BOOLEAN_UPPER:
+  case format_specifiert::STRING_UPPER:
+  case format_specifiert::HASHCODE_UPPER:
+  {
+    string_constraint_generatort::format_specifiert fs_lower=fs;
+    fs_lower.conversion=tolower(fs.conversion);
+    string_exprt lower_case=add_axioms_for_format_specifier(
+      fs_lower, arg, ref_type);
+    return add_axioms_for_to_upper_case(lower_case);
+  }
+  case format_specifiert::DATE_TIME:
+  default:
+    // TODO: printDateTime is not implemented
+    throw std::string("invalid format specifier:").append(1, fs.conversion);
```

Comment:
Use `error()` for the message, `INVARIANT(false, <msg>)` for an invariant, and then `throw 0` so the compiler knows the function exits (this should hopefully change in the future).